posthog/temporal/batch_exports/squash_person_overrides.py

```diff
@@ -468,131 +260,172 @@ async def squash_events_partition(inputs: QueryInputs) -> None:
+    async with heartbeat_every():
+        async with get_client(mutations_sync=2) as clickhouse_client:
+            await clickhouse_client.execute_query(
+                CREATE_JOIN_TABLE_FOR_DELETES_QUERY.format(
+                    database=settings.CLICKHOUSE_DATABASE, dictionary_name=inputs.dictionary_name
+                ),
+            )
 
-    We cannot use the Django ORM in an async context without enabling unsafe behavior.
-    This may be a good excuse to unshackle ourselves from the ORM.
-    """
+            try:
+                await clickhouse_client.execute_query(
+                    DELETE_SQUASHED_PERSON_OVERRIDES_QUERY.format(
+                        database=settings.CLICKHOUSE_DATABASE, dictionary_name=inputs.dictionary_name
+                    ),
+                    query_parameters={
+                        "partition_ids": inputs.partition_ids,
+                        "grace_period": inputs.delete_grace_period_seconds,
+                    },
+                )
 
-    from django.conf import settings
+            finally:
+                await clickhouse_client.execute_query(
+                    DROP_JOIN_TABLE_FOR_DELETES_QUERY.format(database=settings.CLICKHOUSE_DATABASE),
+                )
 
-    activity.logger.info("Deleting squashed persons from Postgres")
-    with psycopg2.connect(
-        dbname=settings.DATABASES["default"]["NAME"],
-        user=settings.DATABASES["default"]["USER"],
-        password=settings.DATABASES["default"]["PASSWORD"],
-        host=settings.DATABASES["default"]["HOST"],
-        port=settings.DATABASES["default"]["PORT"],
-        **settings.DATABASES["default"].get("SSL_OPTIONS", {}),
-    ) as connection:
-        overrides_manager = FlatPostgresPersonOverridesManager(connection)
-        for person_override_to_delete in inputs.iter_person_overides_to_delete():
-            activity.logger.debug("%s", person_override_to_delete)
-            overrides_manager.delete(person_override_to_delete, inputs.dry_run)
+    activity.logger.info("Deleted squashed persons from ClickHouse")
 
 
 @contextlib.asynccontextmanager
-async def person_overrides_dictionary(
-    workflow, query_inputs: QueryInputs, retry_policy: RetryPolicy
-) -> AsyncIterator[str]:
-    """This context manager manages the person_overrides DICTIONARY used during a squash job.
+async def person_overrides_dictionary(workflow, query_inputs: QueryInputs) -> collections.abc.AsyncIterator[None]:
+    """This context manager manages a dictionary used during a squash workflow.
 
-    Managing the DICTIONARY involves setup activities:
-    - Prepare the underlying person_overrides table optimizing the table to remove any duplicates.
-    - Creating the DICTIONARY itself, returning latest_created_at.
+    Managing the dictionary involves setup activities necessary to ensure accurate values land in the
+    dictionary:
+    - Optimizing the table to remove any duplicates.
```

Comment:
I'd like to think that clickhouse is smart enough to figure out that 1) the table is a merge tree, 2) the dictionary keys are the same as the table sort key, 3) thus, any duplicates found are due to unmerged parts, and should be accounted for with a group by using the designated version column to aggregate the latest.

However, I can't find any references to confirm this is indeed happening. So, it could be better to just add the group by in the query ourselves for peace of mind.

Reply:
I've added it, let's see